File: leetcode/Hard/python/25. Reverse Nodes in k-Group.py
'''
https://leetcode.com/problems/reverse-nodes-in-k-group/description/
'''

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    # def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
    def reverseKGroup(self, head, k):
        """
        Разворачивает элементы связанного списка группами по k элементов.
        """
        # Проверка на наличие k элементов
        node = head
        count = 0
        while node and count < k:
            node = node.next
            count += 1
        if count < k:
            return head  # меньше k элементов, возвращаем оригинальный список

        # Разворот текущей группы
        prev, curr = None, head
        for _ in range(k):
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node

        # Рекурсивный вызов для оставшейся части списка
        head.next = self.reverseKGroup(curr, k)
        return prev

''' Полезные ссылки: '''
# 1. 💠Telegram💠❃ Хижина программиста Æ: https://t.me/hut_programmer_07
# 2. 💠Telegram №1💠 @quadd4rv1n7
# 3. 💠Telegram №2💠 @dupley_maxim_1999
# 4. Rutube канал: https://rutube.ru/channel/4218729/
# 5. Plvideo канал: https://plvideo.ru/channel/AUPv_p1r5AQJ
# 6. YouTube канал: https://www.youtube.com/@it-coders
# 7. ВК группа: https://vk.com/science_geeks

File: leetcode/Hard/java/25. Reverse Nodes in k-Group.java
/**
 * https://leetcode.com/problems/reverse-nodes-in-k-group/description/
 */

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

/**
 * Разворачивает элементы связанного списка группами по k элементов.
 */
class Solution {
    /**
     * @param head Голова связанного списка.
     * @param k Размер группы для разворота.
     * @return ListNode Голова нового списка после разворота.
     */
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode node = head;
        int count = 0;
        while (node != null && count < k) {
            node = node.next;
            count++;
        }
        if (count < k) return head; // элементов меньше k

        // Разворот текущей группы
        ListNode prev = null;
        ListNode curr = head;
        for (int i = 0; i < k; i++) {
            ListNode nextNode = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextNode;
        }

        // Рекурсивный вызов для оставшейся части списка
        head.next = reverseKGroup(curr, k);
        return prev;
    }
}

/*
''' Полезные ссылки: '''
# 1. 💠Telegram💠❃ Хижина программиста Æ: https://t.me/hut_programmer_07
# 2. 💠Telegram №1💠 @quadd4rv1n7
# 3. 💠Telegram №2💠 @dupley_maxim_1999
# 4. Rutube канал: https://rutube.ru/channel/4218729/
# 5. Plvideo канал: https://plvideo.ru/channel/AUPv_p1r5AQJ
# 6. YouTube канал: https://www.youtube.com/@it-coders
# 7. ВК группа: https://vk.com/science_geeks
*/